src/Enzo/enzo_EnzoConstrainedTransport.cpp

```diff
@@ -10,30 +10,289 @@
+  //    scheme from Athena++, which requires knowledge of the reconstructed
+  //    densities.
+
+  if ((dim < 0) || (dim > 2)){
+    ERROR("EnzoConstrainedTransport::identify_upwind",
+          "dim has an invalid value");
+  } else {
+    EnzoFieldArrayFactory array_factory(block_, stale_depth);
+    EFlt3DArray density_flux = flux_map.get("density", stale_depth);
+
+    CSlice stale_slc = (stale_depth > 0) ?
+      CSlice(stale_depth,-stale_depth) : CSlice(nullptr, nullptr);
+
+    EFlt3DArray weight_field = weight_l_[dim].subarray(stale_slc, stale_slc,
+                                                       stale_slc);
+
+    // Iteration limits compatible with both 2D and 3D grids
+    for (int iz=0; iz<density_flux.shape(0); iz++) {
+      for (int iy=0; iy<density_flux.shape(1); iy++) {
+        for (int ix=0; ix<density_flux.shape(2); ix++) {
+          // density flux is the face-centered density times the face-centered
+          // velocity along dim
+
+          if ( density_flux(iz,iy,ix) > 0){
+            weight_field(iz,iy,ix) = 1.0;
+          } else if ( density_flux(iz,iy,ix) < 0){
+            weight_field(iz,iy,ix) = 0.0;
+          } else {
+            weight_field(iz,iy,ix) = 0.5;
+          }
+        }
+      }
+    }
+
+  }
+}
+
+//----------------------------------------------------------------------
+
+void EnzoConstrainedTransport::update_all_bfield_components
+(EnzoEFltArrayMap &cur_prim_map, EnzoEFltArrayMap &xflux_map,
+ EnzoEFltArrayMap &yflux_map, EnzoEFltArrayMap &zflux_map,
+ EnzoEFltArrayMap &out_centered_bfield_map, enzo_float dt, int stale_depth)
+{
+  // The current interface values of the interface magnetic fields that are to
+  // be updated are always the values from the start of the timestep
+  std::array<EFlt3DArray,3> *cur_bfieldi_l = &bfieldi_l_;
+  // Now determine the set of arrays where the updated interface bfields are to
+  // be stored:
+  std::array<EFlt3DArray,3> *out_bfieldi_l;
+  if ((partial_timestep_index_== 1) || (num_partial_timesteps_ == 1)){
+    out_bfieldi_l = &bfieldi_l_; // the same as cur_bfieldi_l
+  } else if (partial_timestep_index_== 0){
+    out_bfieldi_l = &temp_bfieldi_l_;
+  } else {
+    ERROR1("EnzoConstrainedTransport::correct_reconstructed_bfield",
+	   "Unsure how to handle current partial_timestep_index_ of %d",
+	   partial_timestep_index_);
+  }
+
+  // First, compute the edge-centered Electric fields (each time, it uses
+  // the current integrable quantities)
+  EnzoConstrainedTransport::compute_all_edge_efields(cur_prim_map, xflux_map,
+                                                     yflux_map, zflux_map,
+                                                     center_efield_,
+                                                     edge_efield_l_, weight_l_,
+                                                     stale_depth);
+
+  EnzoBlock *enzo_block = enzo::block(block_);
+  std::array<enzo_float,3> cell_widths;
```

Comment:
Why is this temp variable needed (versus directly using ` enzo_block->CellWidth`)?